core/src/main/java/org/apache/iceberg/util/PartitionUtil.java

```diff
@@ -51,6 +56,20 @@ private PartitionUtil() {
         MetadataColumns.FILE_PATH.fieldId(),
         convertConstant.apply(Types.StringType.get(), task.file().path()));
 
+    // add _spec_id
+    idToConstant.put(
+        MetadataColumns.SPEC_ID.fieldId(),
+        convertConstant.apply(Types.IntegerType.get(), task.file().specId()));
+
+    // add _partition
+    if (partitionType != null && partitionType.fields().size() > 0) {
+      StructLike coercedPartition = coercePartition(partitionType, spec, partitionData);
+      idToConstant.put(MetadataColumns.PARTITION_COLUMN_ID, convertConstant.apply(partitionType, coercedPartition));
```

Comment:
This needs to call `convertConstant` on each partition value. The assumption in all of the implementations is that it is called on primitive values, not on structs.

Reply:
Okay, I see what's happening. This PR updates the conversion for Spark and because most callers only use `constantsMap(task, func)` there's no way to get a partition passed through. I'm a little uneasy about this, but since we need to convert to a specific class for the reader I don't see a good way around it.

Reply:
Yeah, it is also related to unknown transforms as well.